File: db/user.db.go
package db

import (
	"github.com/Tsuzat/zipit-go-fiber/config"
	"github.com/Tsuzat/zipit-go-fiber/models"
	"github.com/gofiber/fiber/v2/log"
)

func GetUserByEmail(email string) (*models.User, error) {
	user := &models.User{
		Email: email,
	}
	if err := config.DB.Model(user).Where("email = ?", email).Select(); err != nil {
		return nil, err
	}
	if user.Id == "" {
		return nil, models.UserNotFound
	}
	return user, nil
}

func InsertUser(user *models.User) error {
	_, err := config.DB.Model(user).Insert()
	if err != nil || user.Id == "" {
		return err
	}
	log.Info("Created User with id: ", user.Id)
	return nil
}

func UpdateUser(user *models.User) error {
	_, err := config.DB.Model(user).Where("id = ?", user.Id).Update()
	if err != nil || user.Id == "" {
		return err
	}
	log.Info("Updated User with id: ", user.Id)
	return nil
}

func GetUserByIdNameAndEmail(id, name, email string) (*models.User, error) {
	user := &models.User{
		Id:    id,
		Name:  name,
		Email: email,
	}
	if err := config.DB.Model(user).WherePK().Select(); err != nil {
		return nil, err
	}
	return user, nil
}

File: src/database/user.db.ts
import { error } from "console";
import { User } from "../models/user.model";
import db from "./db";
import { users } from "./schemas";
import { eq, and } from "drizzle-orm";

async function insertUser(user: User): Promise<number> {
  const result = await db
    .insert(users)
    .values(user)
    .returning({ id: users.id });
  if (result.length === 0) {
    return -1;
  }
  return result[0].id;
}

type FindUserParams = {
  id?: number;
  email?: string;
};

/**
 * Function to find a user by either ID or email
 * @param params Find user parameters - either 'id' or 'email' must be provided
 * @returns Returns the user object if found, otherwise returns null
 */
async function findUser(params: FindUserParams): Promise<User | null> {
  const { id, email } = params;

  // Ensure at least one parameter is provided
  if (!id && !email) {
    throw new Error("Either 'id' or 'email' must be provided.");
  }

  // Build the where clause dynamically
  const conditions = [];
  if (id) conditions.push(eq(users.id, id));
  if (email) conditions.push(eq(users.email, email));

  const query = db
    .select()
    .from(users)
    .where(and(...conditions)) // Combine conditions with AND
    .limit(1); // Limit the query to one user

  const [user] = await query; // Destructure to get the single user
  return user || null; // Return null if no user found
}

async function updateUser(user: User): Promise<boolean> {
  const result = await db
    .update(users)
    .set(user)
    .where(eq(users.id, user.id))
    .returning({ id: users.id });
  if (result.length === 0) {
    return false;
  }
  return true;
}

export { insertUser, findUser, updateUser };
